File: src/main/java/org/micoli/php/ui/popup/NavigableListCellRenderer.java
package org.micoli.php.ui.popup;

import java.awt.*;
import javax.swing.*;
import org.jetbrains.annotations.NotNull;

public class NavigableListCellRenderer extends DefaultListCellRenderer {

    @Override
    public Component getListCellRendererComponent(
            JList<?> list, Object value, int index, boolean isSelected, boolean cellHasFocus) {

        if (value instanceof NavigableListPopupItem item) {
            return getNavigableItemLabel(list, index, isSelected, cellHasFocus, item);
        }

        return super.getListCellRendererComponent(list, value, index, isSelected, cellHasFocus);
    }

    private @NotNull JLabel getNavigableItemLabel(
            JList<?> list, int index, boolean isSelected, boolean cellHasFocus, NavigableListPopupItem item) {
        if (item instanceof NavigableOpenSearchAction || item instanceof NavigableOpenAllAction) {
            JLabel label = (JLabel) super.getListCellRendererComponent(list, 0, index, isSelected, cellHasFocus);

            label.setText(item.getText());

            return label;
        }

        if (item instanceof NavigableItem navigableItem) {
            FileExtract fileExtract = navigableItem.getFileExtract();
            JLabel label = (JLabel)
                    super.getListCellRendererComponent(list, fileExtract.lineNumber(), index, isSelected, cellHasFocus);

            label.setIcon(navigableItem.getIcon());
            label.setText(navigableItem.getText());

            return label;
        }
        throw new IllegalArgumentException("Unsupported item type");
    }
}

File: src/main/kotlin/org/micoli/php/ui/popup/NavigableListCellRenderer.kt
package org.micoli.php.ui.popup

import java.awt.Component
import javax.swing.DefaultListCellRenderer
import javax.swing.JLabel
import javax.swing.JList

class NavigableListCellRenderer : DefaultListCellRenderer() {
    override fun getListCellRendererComponent(
        list: JList<*>?,
        value: Any?,
        index: Int,
        isSelected: Boolean,
        cellHasFocus: Boolean,
    ): Component? {
        if (value is NavigableListPopupItem) {
            return getNavigableItemLabel(list, index, isSelected, cellHasFocus, value)
        }

        return super.getListCellRendererComponent(list, value, index, isSelected, cellHasFocus)
    }

    private fun getNavigableItemLabel(
        list: JList<*>?,
        index: Int,
        isSelected: Boolean,
        cellHasFocus: Boolean,
        item: NavigableListPopupItem?,
    ): JLabel {
        if (item is NavigableOpenSearchAction || item is NavigableOpenAllAction) {
            val label =
                super.getListCellRendererComponent(list, 0, index, isSelected, cellHasFocus)
                    as JLabel

            label.setText(item.getText())

            return label
        }
        if (item is NavigableItem) {
            val fileExtract = item.fileExtract
            val label =
                super.getListCellRendererComponent(
                    list, fileExtract.lineNumber, index, isSelected, cellHasFocus) as JLabel

            label.setIcon(item.icon)
            label.setText(item.getText())

            return label
        }
        throw IllegalArgumentException("Unsupported item type")
    }
}
